Pandas Data Frames
==========

The DataFrame is the way Pandas represents a table, and Series is the data-structure Pandas use to represent a column. So, a data frame and series are synonomous with table and column.

What makes Pandas so attractive is the powerful interface to access individual records of the table, proper handling of missing values, and relational-databases operations between DataFrames.

In [17]:
# LOAD IN A DATA FRAME
from matplotlib import pyplot as plt
import pandas as pd
df = pd.read_csv('../data/gapminder_gdp_europe.csv', index_col='country')

## Inspecting Data
We've already see how we can get information about a dataframe using the `.info()` and `.describe()` functions, but there are many ways to get information and view a data frame

* Can also use describe() on data frame selections (like a single column)

In [13]:
print( df["gdpPercap_1982"].describe() )

count       30.000000
mean     15617.896551
std       6453.234827
min       3630.880722
25%      11449.870115
50%      15322.824720
75%      20901.729730
max      28397.715120
Name: gdpPercap_1982, dtype: float64


* We can print the first or last x number of rows of our data frame using the head() and tails() functions.

In [14]:
print(df.head(3))

         gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  gdpPercap_1967  \
country                                                                   
Albania     1601.056136     1942.284244     2312.888958     2760.196931   
Austria     6137.076492     8842.598030    10750.721110    12834.602400   
Belgium     8343.105127     9714.960623    10991.206760    13149.041190   

         gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
country                                                                   
Albania     3313.422188      3533.00391     3630.880722     3738.932735   
Austria    16661.625600     19749.42230    21597.083620    23687.826070   
Belgium    16672.143560     19117.97448    20979.845890    22525.563080   

         gdpPercap_1992  gdpPercap_1997  gdpPercap_2002  gdpPercap_2007  
country                                                                  
Albania     2497.437901     3193.054604     4604.211737     5937.029526  
Austria    27042.018680   

In [15]:
print(df.tail(3))

                gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
country                                                          
Switzerland       14734.232750    17909.489730    20431.092700   
Turkey             1969.100980     2218.754257     2322.869908   
United Kingdom     9979.508487    11283.177950    12477.177070   

                gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  \
country                                                          
Switzerland       22966.144320     27195.11304    26982.290520   
Turkey             2826.356387      3450.69638     4269.122326   
United Kingdom    14142.850890     15895.11641    17428.748460   

                gdpPercap_1982  gdpPercap_1987  gdpPercap_1992  \
country                                                          
Switzerland       28397.715120    30281.704590    31871.530300   
Turkey             4241.356344     5089.043686     5678.348271   
United Kingdom    18232.424520    21664.787670    22705.092540   

       

In [11]:
print(df.dtypes)

gdpPercap_1952    float64
gdpPercap_1957    float64
gdpPercap_1962    float64
gdpPercap_1967    float64
gdpPercap_1972    float64
gdpPercap_1977    float64
gdpPercap_1982    float64
gdpPercap_1987    float64
gdpPercap_1992    float64
gdpPercap_1997    float64
gdpPercap_2002    float64
gdpPercap_2007    float64
dtype: object


* Use `shape` to get the row and column numbers

In [16]:
print(df.shape)

(30, 12)


Here we can see the that the data have 30 rows of data and 12 attributes worth of information.

* Use the `len()` function to get numbers of each individually

In [ ]:
# print number of rows of data
print(len(df))

In [ ]:
# print number of columns of data
print(len(df.columns))

* Get the column names with `columns()`

In [ ]:
print(df.columns)

* Use a column name to get all values for that columns

---
## EXERCISE:
1. How many countries are there in the gapminder_all.csv file?
2. What is the last country listed in the table?
---

---
## Get information about a particular column

* Operations like mean, max, min, can be used on individual columns

In [22]:
# Mean GDP in 1967
print(df["gdpPercap_1967"].mean())
# Mean GDP in 1972
print(df["gdpPercap_1972"].mean())
# Mean GDP in 1977
print(df["gdpPercap_1977"].mean())


10143.823756533333
12479.57524646667
14283.9791096


---
## EXERCISE:
1. What is the average (mean) GDP value for all countries in 1992?
2. What about the max value for all countries in 1952?


---

## Rearange Columns

* Difficult to do using a csv library or by hand
* The reverse() function will reverse the ordering of a list
     * E.g.   `['a', 'b', 'c']` to `['c', 'b', 'a']`

In [24]:
cols = list(df.columns)
print( cols )

cols.reverse()
print ( cols )

['gdpPercap_1952', 'gdpPercap_1957', 'gdpPercap_1962', 'gdpPercap_1967', 'gdpPercap_1972', 'gdpPercap_1977', 'gdpPercap_1982', 'gdpPercap_1987', 'gdpPercap_1992', 'gdpPercap_1997', 'gdpPercap_2002', 'gdpPercap_2007']
['gdpPercap_2007', 'gdpPercap_2002', 'gdpPercap_1997', 'gdpPercap_1992', 'gdpPercap_1987', 'gdpPercap_1982', 'gdpPercap_1977', 'gdpPercap_1972', 'gdpPercap_1967', 'gdpPercap_1962', 'gdpPercap_1957', 'gdpPercap_1952']


* Using that now reversed list above, we can create a new list, with the values order in reverse

In [ ]:
new_df = df[cols]
new_df.head(3)

## Transposing tables

In many cases we may need to transpose the column and rows in a table.  Pandas allows us to to this easily with the `T` command.


In [10]:
# Print first three rows of the data frame
print("Data Frame:\n", df.head(3))
# Transpose the dataframe and print the first three rows
print("\n\nTransposed:\n",df.T.head(3))

Data Frame:
          gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  gdpPercap_1967  \
country                                                                   
Albania     1601.056136     1942.284244     2312.888958     2760.196931   
Austria     6137.076492     8842.598030    10750.721110    12834.602400   
Belgium     8343.105127     9714.960623    10991.206760    13149.041190   

         gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
country                                                                   
Albania     3313.422188      3533.00391     3630.880722     3738.932735   
Austria    16661.625600     19749.42230    21597.083620    23687.826070   
Belgium    16672.143560     19117.97448    20979.845890    22525.563080   

         gdpPercap_1992  gdpPercap_1997  gdpPercap_2002  gdpPercap_2007  
country                                                                  
Albania     2497.437901     3193.054604     4604.211737     5937.029526  
Austria    27

---
## EXERCISE:
1. Read in a new data frame for the gapminder_gdp_americas.csv file
2. Print the last last three columns of the data frame

---

## Selecting values

Remember that a DataFrame provides a index as a way to identify the rows of the table. A row also has a position inside the table as well as a label, which uniquely identifies its entry in the DataFrame.

To access a value at the position [ i , j ] (row, column) of a DataFrame, we have two options, depending on what is the meaning of i in use.



We can see that the first value in the first column is `1601.056136`. Remember that in programming, we begin our counting at zero, so this index would be `0,0`, not `1,1`.

### Use DataFrame.iloc[..., ...] to select values by their position
* Allows you to specify location by numerical index similar to 2D version of character selection in strings.


In [ ]:
print("\nData value in first row at first column: ", df.iloc[0, 0])

In [ ]:
print("\nData value in fifth row at third column: ", df.iloc[4, 2])

### Use `DataFrame.loc[..., ...]` to select values by their (entry) label.

*   Can specify location by name or by numerical index.

In [ ]:
print(df.loc["Albania", "gdpPercap_1952"])

In [ ]:
print(df.loc['Bulgaria', "gdpPercap_1962"])

---
## EXERCISE
~~~
import pandas
df = pandas.read_csv('data/gapminder_gdp_europe.csv', index_col='country')
~~~

1. Write an expression to find the Per Capita GDPs for Serbia (Serbia is in the Europe CSV data file).
1. Write an expression to find the Per Capita GDP for Serbia in 2007.

---

### Use `:` on its own to mean all columns or all rows.

*   Just like Python's usual slicing notation, we can print all columns or all rows with `.loc` using the `:`

In [ ]:
print(df.loc["Albania",:])

* Would get the same result printing `df.iloc[0]` (without a second index).
* We can also omit the `:` and get the same result in either case.
    * e.g. `df.loc["Albania"]`

In [ ]:
print(df.loc[:, "gdpPercap_1952"])

*   Would get the same result printing `df["gdpPercap_1952"]`
*   Also get the same result printing `df.gdpPercap_1952` (since it's a column name)

---
## EXERCISE:
1. Print out all of the country gdp values for 1972

---

### We can also use the `:` to select sections of a table 
* Similar to the way we would select a section of from a normal python list, we can do the same with data frames.

In [ ]:
print(df.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972'])

In the above code, we're selecting all row from Italy to Poland and all years from 1962 tto 1972.

Note that in Pandas **slicing using indexes is inclusive at both ends**, which differs from typical python behavior where slicing indicates everything up to but not including the final index.

### Select multiple columns or rows using `DataFrame.iloc` and a named slice.
* We can also make selection from a data frame using the index location of the row or column
    * Remember that in programming languages, we start counting at 0

In [34]:
print("First row of data frame:\n", df.head(1))

print("\n\nValue in the first row of the first column:\n",df.iloc[0,0])
print("Values in the first row of the first two columns:\n", df.iloc[0,0:2])

First row of data frame:
          gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  gdpPercap_1967  \
country                                                                   
Albania     1601.056136     1942.284244     2312.888958     2760.196931   

         gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
country                                                                   
Albania     3313.422188      3533.00391     3630.880722     3738.932735   

         gdpPercap_1992  gdpPercap_1997  gdpPercap_2002  gdpPercap_2007  
country                                                                  
Albania     2497.437901     3193.054604     4604.211737     5937.029526  


Value in the first row of the first column:
 1601.056136
Values in the first row of the first two columns:
 gdpPercap_1952    1601.056136
gdpPercap_1957    1942.284244
Name: Albania, dtype: float64


* **Note that unlike slicing using column or row names, slicing using indexes is not inclusive** 

---
## EXERCISE:
1. Print out all values from Hungary through Montenegro for the years 1977 through 1997

---

---
## EXERCISE:
1.  Do the two statements below produce the same output?
    ~~~
    print(df.iloc[0:2, 0:2])
    print(df.loc['Albania':'Belgium', 'gdpPercap_1952':'gdpPercap_1962'])
    ~~~

1.  Based on this,what rule governs what is included (or not) in numerical slices and named slices in Pandas?
---

### Slicing individual rows and columns
* Instead of creating slices of *this* to *that* using the `:`, we can also slice using individual rows and columns by placing names or indexes in brackets `[]`.

In [41]:
print(df.loc[["Italy", "Austria", "United Kingdom"], ["gdpPercap_2007", "gdpPercap_1957"]])

                gdpPercap_2007  gdpPercap_1957
country                                       
Italy              28569.71970     6248.656232
Austria            36126.49270     8842.598030
United Kingdom     33203.26128    11283.177950


---
## EXERCISE:
1. Using the index locations, `print` out the first, third, and eight columns for the sixteenth through nineteenth rows.

## Result of slicing can be used in further operations.

In [ ]:
print(df.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972'].max())

In [ ]:
print(df.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972'].min())

*   Usually don't just print a slice.
*   All the statistical operators that work on entire data frames work the same way on slices.

## Create data frame from selections

* We can create new data frame by selecting data frames based on values and assigining it to a variable

In [64]:
# Use a subset of data to keep output readable.
subsetdf = df.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972']
print('Subset of data:\n', subsetdf)

Subset of data:
              gdpPercap_1962  gdpPercap_1967  gdpPercap_1972
country                                                    
Italy           8243.582340    10022.401310    12269.273780
Montenegro      4649.593785     5907.850937     7778.414017
Netherlands    12790.849560    15363.251360    18794.745670
Norway         13450.401510    16361.876470    18965.055510
Poland          5338.752143     6557.152776     8006.506993


## Create DataFrame using query
* We can query values in a data frame to create new selections
* By passing a dataframe query to itself, we can create a new dataframe with only those values

In [72]:
query_10k = subsetdf["gdpPercap_1962"] >= 10000
subset10kdf = subsetdf[query]

print(subset10kdf)
print(subset10kdf.shape)

             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972
country                                                    
Netherlands     12790.84956     15363.25136     18794.74567
Norway          13450.40151     16361.87647     18965.05551
(2, 3)


---
## EXERCISE:

* Create three data frames and get the size of each one.
    1. Countries with a gdp per capita in 1952 above 10000
    1. Countries with a gdp per capita in 1962 above 10000
    1. Countries with a gdp per capita in 1972 above 10000

---

## Filter a DataFrame using a Boolean mask

* A frame full of Booleans is sometimes called a *mask* because of how it can be used
* Comparison is applied element by element
* Returns a similarly-shaped data frame of `True` and `False`

In [73]:
mask10k = subsetdf >= 10000
print( mask10k )

            gdpPercap_1962 gdpPercap_1967 gdpPercap_1972
country                                                 
Italy                False           True           True
Montenegro           False          False          False
Netherlands           True           True           True
Norway                True           True           True
Poland               False          False          False


* We can use masks to filter an entire dataframe with a single query
    * More efficient than using a single query on multiple columns

In [87]:
subset = subsetdf[mask10k]
print(subset)
print("Shape: ", subset.shape)

             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972
country                                                    
Italy                   NaN     10022.40131     12269.27378
Montenegro              NaN             NaN             NaN
Netherlands     12790.84956     15363.25136     18794.74567
Norway          13450.40151     16361.87647     18965.05551
Poland                  NaN             NaN             NaN
Shape:  (5, 3)


*   Get the value where the mask is true, and NaN (Not a Number) where it is false.
*   Useful because NaNs are ignored by operations like max, min, average, etc.


* If we wanted to remove all rows with a NaN value in any column we could use the `.dropna()` function

In [88]:
print(subset.dropna())
print(subset.dropna().shape)

             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972
country                                                    
Netherlands     12790.84956     15363.25136     18794.74567
Norway          13450.40151     16361.87647     18965.05551
(2, 3)


## Create new columns

* We can easily create new columns in the same way we would add a key and value to a dictionary

In [ ]:
# Create a new column diff_07_52 that is the difference between gdp per capita from 1952 to 2007
df["diff_07_52"] = df["gdpPercap_2007"] - df["gdpPercap_1952"]
df.head()

> ## Reconstructing Data
>
> Explain what each line in the following short program does:
> what is in `first`, `second`, etc.?
>
> ~~~
> first = pandas.read_csv('data/gapminder_gdp_all.csv', index_col='country')
> second = df[df['continent'] == 'Americas']
> third = second.drop('Puerto Rico')
> fourth = third.drop('continent', axis = 1)
> fourth.to_csv('result.csv')
> ~~~

>>### Solution
>>Let’s go through this piece of code line by line.
>>
>>`first = pandas.read_csv('data/gapminder_all.csv', index_col='country')`
>>
>>This line loads the dataset containing the GDP data from all countries into a dataframe called first. The index_col='country' parameter selects which column to use as the row labels in the dataframe.
>>
>>`second = first[first['continent'] == 'Americas']`
>>
>>This line makes a selection: only those rows of first for which the ‘continent’ column matches ‘Americas’ are extracted. Notice how the Boolean expression inside the brackets, first['continent'] == 'Americas', is used to select only those rows where the expression is true. Try printing this expression! Can you print also its individual True/False elements? (hint: first assign the expression to a variable)
>>
>>`third = second.drop('Puerto Rico')`
>>
>>As the syntax suggests, this line drops the row from second where the label is ‘Puerto Rico’. The resulting dataframe third has one row less than the original dataframe second.
>>
>>`fourth = third.drop('continent', axis = 1)`
>>
>>Again we apply the drop function, but in this case we are dropping not a row but a whole column. To accomplish this, we need to specify also the axis parameter (we want to drop the second column which has index 1).
>>
>>`fourth.to_csv('result.csv')`
>>
>>The final step is to write the data that we have been working on to a csv file. Pandas makes this easy with the to_csv() function. The only required argument to the function is the filename. Note that the file will be written in the directory from which you started the Jupyter or Python session.

## EXERCISE:
1. Explain in simple terms what `idxmin` and `idxmax` do in the short program below.
    ~~~
    df = pd.read_csv('data/gapminder_gdp_europe.csv', index_col='country')
    print(df.idxmin())
    print(df.idymax())
    ~~~

2. When would you use these methods?

In [128]:
df = pd.read_csv('../data/gapminder_gdp_europe.csv', index_col='country')
print(df.idxmin())
print(df.idxmax())

gdpPercap_1952    Bosnia and Herzegovina
gdpPercap_1957    Bosnia and Herzegovina
gdpPercap_1962    Bosnia and Herzegovina
gdpPercap_1967    Bosnia and Herzegovina
gdpPercap_1972    Bosnia and Herzegovina
gdpPercap_1977    Bosnia and Herzegovina
gdpPercap_1982                   Albania
gdpPercap_1987                   Albania
gdpPercap_1992                   Albania
gdpPercap_1997                   Albania
gdpPercap_2002                   Albania
gdpPercap_2007                   Albania
dtype: object
gdpPercap_1952    Switzerland
gdpPercap_1957    Switzerland
gdpPercap_1962    Switzerland
gdpPercap_1967    Switzerland
gdpPercap_1972    Switzerland
gdpPercap_1977    Switzerland
gdpPercap_1982    Switzerland
gdpPercap_1987         Norway
gdpPercap_1992         Norway
gdpPercap_1997         Norway
gdpPercap_2002         Norway
gdpPercap_2007         Norway
dtype: object


---
## PRACTICE EXERCISE.
Using the Gapminder GDP data for Europe, write an expression to select each of the following:
1.  GDP per capita for all countries in 1982.
1.  GDP per capita for Denmark for all years.
1.  GDP per capita for all countries for years *after* 1985.
1.  GDP per capita for each country in 2007 as a multiple of GDP per capita for that country in 1952.
---

# -- COMMIT YOUR WORK TO GITHUB --

---
## Keypoints:
 - "Use `DataFrame.iloc[..., ...]` to select values by index location."
 - "Use `:` on its own to mean all columns or all rows."
 - "Select multiple columns or rows using `DataFrame.ix` and a named slice."
 - "Result of slicing can be used in further operations."
 - "Use comparisons to select data based on value."
 - "Select values or NaN using a Boolean mask."